### PR TITLE
Add retry logic and proxy support to the NeMo LLM Service

### PR DESCRIPTION
**Pull request by mdemoret-nv**
## Description

- Adds the ability to retry NeMo failures more than one time
- Adds an argument to configure the retry count
- Adds support for proxying requests to the NeMo service using the `NGC_API_BASE` environment variable
- Changes the API for the base `LLMService` to improve the type hints and allow arguments other than strings to be used.

## By Submitting this PR I confirm:
- I am familiar with the [Contributing Guidelines](https://github.com/nv-morpheus/Morpheus/blob/main/docs/source/developer_guide/contributing.md).
- When the PR is ready for review, new or existing tests cover these changes.
- When the PR is ready for review, the documentation is up to date with these changes.

### CODE REVIEW

File: morpheus/llm/services/nemo_llm_service.py [hunk 1]
```diff
@@ -114,22 +114,34 @@ async def generate_batch_async(self, inputs: dict[str, list[str]]) -> list[str]:
...
 
         for result in results:
-            result = nemollm.NemoLLM.post_process_generate_response(result, return_text_completion_only=False)
-            if result.get('status', None) == 'fail':
-                raise RuntimeError(result.get('msg', 'Unknown error'))
-
-            responses.append(result['text'])
+            responses.append(result)
 
         return responses
```
Comment (dagardner-nv) [inline morpheus/llm/services/nemo_llm_service.py:146]: If we aren't doing anything with the results why not just return results and drop this loop?

File: morpheus/llm/services/nemo_llm_service.py [hunk 2]
```diff
@@ -114,22 +114,34 @@ async def generate_batch_async(self, inputs: dict[str, list[str]]) -> list[str]:
...
         """
         prompts = inputs[self._prompt_key]
-        futures = [
-            asyncio.wrap_future(
-                self._parent._conn.generate(self._model_name, p, return_type="async", **self._model_kwargs))
-            for p in prompts
-        ]
 
-        results = await asyncio.gather(*futures)
+        async def process_one(p: str):
```
Comment (dagardner-nv) [inline morpheus/llm/services/nemo_llm_service.py:118]: I would make this a method on the class and not nested.

File: morpheus/llm/services/nemo_llm_service.py [hunk 3]
```diff
@@ -86,79 +87,138 @@ async def generate_async(self, input_dict: dict[str, str]) -> str:
...
-        responses = []
-
-        for result in results:
-            result = nemollm.NemoLLM.post_process_generate_response(result, return_text_completion_only=False)
-            if result.get('status', None) == 'fail':
-                raise RuntimeError(result.get('msg', 'Unknown error'))
+        futures = [self._process_one_async(p) for p in prompts]
 
-            responses.append(result['text'])
+        results = await asyncio.gather(*futures, return_exceptions=False)
```
Comment (dagardner-nv) [inline morpheus/llm/services/nemo_llm_service.py:183]: Shouldn't this be?
```
results = await asyncio.gather(*futures, return_exceptions=return_exceptions)
```

File: morpheus/llm/services/openai_chat_service.py [hunk 4]
```diff
@@ -231,18 +264,37 @@ def generate_batch(self, inputs: dict[str, list[str]]) -> list[str]:
         results = []
         for (i, prompt) in enumerate(prompts):
             assistant = assistants[i] if assistants is not None else None
-            results.append(self._generate(prompt, assistant))
+            if (return_exceptions):
+                results.append(self._generate(prompt, assistant, return_exceptions=True))
+            else:
+                results.append(self._generate(prompt, assistant, return_exceptions=False))
```
Comment (dagardner-nv) [inline morpheus/llm/services/openai_chat_service.py:270]: Is the if statement needed here? Couldn't it just be:
```
results.append(self._generate(prompt, assistant, return_exceptions=return_exceptions))
```

**Review by dagardner-nv — APPROVED** [review]
Just a few questions

File: morpheus/llm/services/openai_chat_service.py (hunk 4 above)
Comment (mdemoret-nv) [reply morpheus/llm/services/openai_chat_service.py:270]: This allows the type hinting to actually work and prevents pylint errors all over the place

File: morpheus/llm/services/nemo_llm_service.py (hunk 3 above)
Comment (mdemoret-nv) [reply morpheus/llm/services/nemo_llm_service.py:183]: Good catch. Fixed

**Pull request merged**